### ISSUE TITLE
sdbus++ support for async method calls

### ISSUE THREAD

**Issue by charles-stracener** [openbmc/sdbusplus#72]
@williamspatrick sdbusplus currently has async client call support (thank you for that), and I was wondering if there are any plans to add an async flag to sdbus++ methods?  I have a fairly simple solution implemented and working, but I didn't want to interfere with any existing plans.

I created an sdbus++ method flag called 'deferred', which causes the code generation to defer the message to the service's method implementation, and in this state the server.cpp callback does not send a message reply.  Instead, the service method implementation now issues the sdbusplus::message reply once it has completed its work.

This works well with an event loop, non-blocking, actor model.  If this is something that you think could benefit the project, then I can create a pull request for feedback.  If there's already existing plans for something similar, I'd love to hear more about it.

**Comment by williamspatrick** [comment]
I don't think adding it as a flag is too helpful.  That means that all implementations of that function would need to do it in that async way.  If we went that route I think we'd end up just making everything async that way.

I've been working in the background, as I have time, at adding support C++20 co-routines.  My intention is generate an async version of the files once I have the co-routine support complete.  Using C++20 concepts I can actually detect the return value from the implementation to differentiate between `task<return_value>` (async) and `return_value` (sync).

**Comment by superchalupa** [comment]
Hey, Patrick, if Danny posted something with the C++ concepts approach, would that be helpful? What's your outlook on completing your approach? We've got a couple of projects that are blocked on this support and Danny can help if you're ok with that.

**Comment by superchalupa** [comment]
How much work is the coroutine approach vs just deferring back to the event loop? Is there a way to do the simplistic return value check and return to the event loop without coroutines and add that later? I was thinking (without knowing as much in depth about the architecture here as I need to) that the code to detect return value is already in the asio directory, and just adding a little of that to return to the event loop would do the trick, where the function writer would be responsible for managing the return message themselves.

The other useful trick here would be around detecting function parameters to send the message as a parameter, optionally, if the method needs it.

**Comment by williamspatrick** [comment]
> sdbusplus currently has async client call support (thank you for that), and I was wondering if there are any plans to add an async flag to sdbus++ methods? 

Keep in mind that you _can_ write servers with those ASIO methods also.  I haven't really worked on ASIO at all so you don't have me to thank for it.  What I don't like about it is that it they are very manual to formulate and they don't give any compile-time guarantees that the interfaces match the YAML.  As part of this co-routine work I hope to also write proper YAML-generated client bindings / proxy objects.

> Hey, Patrick, if Danny posted something with the C++ concepts approach, would that be helpful? What's your outlook on completing your approach? We've got a couple of projects that are blocked on this support and Danny can help if you're ok with that.

I'm a few months out from it being ready, I suspect.  Sure, if someone wants to try to enhance it as-is, that is fine.

> How much work is the coroutine approach vs just deferring back to the event loop? Is there a way to do the simplistic return value check and return to the event loop without coroutines and add that later?

I don't know how to do it with the current API without some serious hacky code.  The current APIs implement a pure virtual function: `method(args...) -> return_value = 0`.  The message isn't there and there is no way to return anything except the `return_value`.  If we change those APIs we'd end up breaking all the current `method(args...) override` implementations.  With the virtual functions, as-is, I don't have a good way to (a) get the message so you could hold it for later reply and (b) communicate to the current bindings that there is no return [yet].

When I do the co-routine implementation I intend to get rid of all the pure-virtual polymorphism because C++ concepts allow you to do so much more at compile time now*.  We can both ensure that the code only compiles if the inheriting class satisfies the appropriate "concepts" (which is equivalent to `virtual`+`override` relationship presently) _and_ do compile time detection of which return type is coming from the member function that satisfies the concept.

(*) All of this probably _could_ have been done back when I originally wrote this in C++14 but it was significantly more work and ugly templates and I never thought we'd need to actively change the APIs like this.  Our original design was that no daemons should talk to other daemons, so async wasn't expected to be necessary.

**Comment by williamspatrick** [comment]
Async server bindings are now implemented.
